---
title: "Multi-omics variable selection using MRF framework"
author: 
  - Wei Zhang^[University of Miami]
date: "`r Sys.Date()`"
output:
  rmarkdown::html_document:
    highlight: pygments
    theme: yeti
    toc: true
    number_sections: true
    df_print: paged
    code_download: true
    toc_float:
      collapsed: yes
    toc_depth: 3
editor_options:
  chunk_output_type: inline 
bibliography: ref.bib
reference_section_title: "References"
---

```{r setup, include=T}
knitr::opts_chunk$set(echo = TRUE)
# Directory
dir_base <- ".."
```

<style>
body {
  line-height: 1.5;
}
</style>

# Load packages and functions

```{r message = F, warning = F}
# Install and load packages
source(file.path(dir_base, "load_package.R"))
script <- list.files( 
  path = file.path(dir_base,"code/function"),
  pattern = "[.]R$", 
  full.names = T,
  recursive = T
)
for (f in script) source(f)
library(InterSIM) # Use for simulation of multi-omics data
```

# Variable selection

Our multi-omics MRF variable selection framework is implemented using the randomforestSRC R package [@RFSRC2024].

## Simulation of three-omics data

To generate omics data, we simulated mRNA, miRNA, and DNA methylation datasets with 150 samples using
the `InterSim()` function InterSim package [@InterSIM2018]. This package allows for the simulation of multiple interrelated data types, including DNA methylation, mRNA gene expression,
and protein expression, based on the TCGA ovarian cancer data. Here we set the
mean cluster shift of all data to 0.6 and 4 clusters are generated with the
proportions of (0.1, 0.3, 0.4, 0.2).

```{r}
set.seed(12354)
sim1 <- InterSIM(n.sample = 150,
                 delta.methyl = .6,
                 cluster.sample.prop=c(0.1,0.3,0.4,0.2),
                 delta.expr = .6,
                 delta.protein = .6)

dat_list <- sim1[1:3]
label <- sim1$clustering.assignment$cluster.id
```

```{r}
# Ensure the input for MRF functions is a named list, where each element in the list is a data frame
class(dat_list) 
names(dat_list)  

# Note: The column names in each data frame must not contain special characters like "."
dat.list <- dat_list %>% purrr::map(~data.frame(.))
dat.list$dat.protein <- dat_list$dat.protein %>% janitor::clean_names()  
```

The dimensions of each data:

```{r}
purrr::map(dat_list, ~dim(.))
```

## Run initial MRF model

Before performing variable selection, the initial MRF model must be fitted using the `mrf3_init()` function. Below is an explanation of the key parameters:

- **`dat.list`**: A list containing multi-omics datasets with samples in columns and features in rows. Samples should be matched across datasets.
- **`ntree`**: Number of trees for fitting the MRF model. Default is 300.
- **`scale`**: Whether to z-standardize each feature. Default is `TRUE`.
- **`yprob`**: Probability of response features being selected in each node split. Default is `0.5`.
- **`connect_list`**: Pre-defined connection list between datasets. If provided, variable selection uses this list. If not, the algorithm finds optimal connections between datasets.
- **`var_prop`**: Proportion of variance explained by PC datasets when finding optimal connections. Default is `0.6`.
- **`direct`**: Logical; determines whether to keep both directions in the connection list for optimal connections.
- **`lambda`**: Penalizes variables selected only once in a tree. Experimental parameter. Default is `1`.
- **`normalized`**: Logical; determines whether to use normalized variable weights. Default is `TRUE`.
- **`use_depth`**: Logical; determines whether to compute the average IMD selected in a tree. Default is `FALSE`.
- **`calc`**: Select which weights to calculate: `"X"`, `"Y"`, or `"Both"`. Use when fewer than two datasets are in the model. Default is `"Both"`.
- **`parallel`**: Logical; determines whether to use parallel computation for weight calculation.
- **`return_data`**: Whether to return the data list. Default is `FALSE`.
- **`cores`**: Number of cores to use for computation.


```{r}
mod <- mrf3_init(dat.list = dat_list)
# This might take a while to run the function.
```

The variable inverse minimal depth (IMD) for data are saved under `weights`. Here we can plot the IMD using the `plot_weights()` function in utility folder.

```{r fig.width=12, fig.height=4}
plot_weights(mod$weights, top = NULL)
```


## Conduct variable selection

The variable selection for different omics data can be conducted using `mrf3_vs()` function simoutanously. The methods described in the manuscript can be selected by the parameter `method`:

- **`"filter"`**: Variable filtering. The adjustable parameter `k` specifies the number of times the out-of-bag (OOB) error is computed. The default value is 3.
- **`"mixture"`**: Detecting signals with mixture model. The adjustable parameters are `c1`, which specifies the prior distribution of the first component (options: "normal" or "truncn"), and `c2`, which specifies the prior distribution of the second component (options: "normal" or "gamma").
- **`"test"`**: IMD transformation. The adjustable parameter `level` specifies the significance level for selecting important variables. The default value is 0.05.

In this example, we used the mixture model method for illustration.

```{r}
mod_vs <- mrf3_vs(
  mod = mod,
  dat.list = dat_list,
  method = "mixture"
)
```

```{r fig.width=12, fig.height=4}
plot_weights(mod_vs$weights, top = NULL)
```

After variable selection, the number of variables reduced to 

```{r}
purrr::map(mod_vs$weights, ~length(.[.>0]))
```

# Clustering analysis using IntNMF (Optional)

To demonstrate how the selected variables improve clustering results, we used the IntNMF method [@IntNMF2018] to perform clustering analysis. We began by conducting the analysis using the original data:

```{r}
# Scale the data and shift all values to be positive.
df <- dat_list %>% purrr::map(~ {
  . <- scale(.)
  if (!all(. > 0)) {
    m <- abs(min(.))
    . <- pmax(. + m, 0)
  }
  as.matrix(./max(.))
})

# Perform clustering with IntNMF
mod_int_org <- IntNMF::nmf.mnnals(df, k = 4)

# Calculate the Adjusted Rand Index (ARI) for the clustering results
ari_org <- mclust::adjustedRandIndex(mod_int_org$clusters, label)
print(ari_org)
```

```{r fig.width=12, fig.height=4}
g1 <- plot_tSNE(dat_list$dat.expr, main = "Original: mRNA", group = factor(label), label_group = F)
g2 <- plot_tSNE(dat_list$dat.methyl, main = "Original: methyl", group = factor(label), label_group = F)
g3 <- plot_tSNE(dat_list$dat.protein, main = "Original: protein", group = factor(label), label_group = F)
ggpubr::ggarrange(g1,g2,g3, nrow = 1, ncol = 3)
```

Next, we performed the clustering analysis using the selected variables:

```{r}
# Scale the data and shift all values to be positive.
df <- mod_vs$dat.list %>% purrr::map(~ {
  . <- scale(.)
  if (!all(. > 0)) {
    m <- abs(min(.))
    . <- pmax(. + m, 0)
  }
  as.matrix(./max(.))
})

# Perform clustering with IntNMF using the selected variables
mod_int_vs <- IntNMF::nmf.mnnals(df, k = 4)

# Calculate the Adjusted Rand Index (ARI) for the clustering results
ari_vs <- mclust::adjustedRandIndex(mod_int_vs$clusters, label)
print(ari_vs)
```

```{r fig.width=12, fig.height=4}
g1 <- plot_tSNE(mod_vs$dat.list$dat.expr, main = "Selected: mRNA", group = factor(label), label_group = F)
g2 <- plot_tSNE(mod_vs$dat.list$dat.methyl, main = "Selected: methyl", group = factor(label), label_group = F)
g3 <- plot_tSNE(mod_vs$dat.list$dat.protein, main = "Selected: protein", group = factor(label), label_group = F)
ggpubr::ggarrange(g1,g2,g3, nrow = 1, ncol = 3)
```

**Results**

We observe that the Adjusted Rand Index (ARI) improves significantly when using the selected variables, indicating better clustering performance.

# Seesion information

<details>
```{r}
devtools::session_info()
```
</details>

# References


